src/System.Linq/src/System/Linq/Enumerable.SelectWithCountIterator.cs

```diff
@@ -0,0 +1,62 @@
+﻿
+using System.Collections.Generic;
+
+namespace System.Linq
+{
+    public static partial class Enumerable
+    {
+        internal abstract class SelectWithCountIterator<TSource> : SelectIterator<TSource>, ICollection<TSource>
```

Comment:
I'd prefere to remove this class and implement ```ICollection<T>``` independently for```SelectArrayIterator```, ```SelectListIterator``` and ```SelectCollectionIterator```. That would be better for performance.